dartagnan/src/main/java/com/dat3m/dartagnan/program/expression/type/TypeFactory.java

```diff
@@ -0,0 +1,55 @@
+package com.dat3m.dartagnan.program.expression.type;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class TypeFactory {
+
+    private static final TypeFactory instance = new TypeFactory();
+    private final BooleanType booleanType = new BooleanType();
+    private final NumberType numberType = new NumberType();
+    private final PointerType pointerType = new PointerType();
+    private final Map<Integer, IntegerType> integerTypeMap = new HashMap<>();
+    private final Map<Type, Map<Integer, ArrayType>> arrayTypeMap = new HashMap<>();
+    private final Map<List<Type>, AggregateType> aggregateTypeMap = new HashMap<>();
```

Comment:
Rather than using the constructor parameters as key, it is easier to use the type object itself as key.
As long as there is only a single constructor with a single parameter, it might be fine, but if that changes, then your approach doesn't work nicely (the nested map for array types is already quite ugly IMO).
Just construct the instance and then use it to check if there is a canonical instance in the map.
Indeed, I would probably introduce a `Canonicalizer` class that does precisely that: it takes an object as input and returns a canonical version of it.